src/connectivity/mod.rs

```diff
@@ -192,6 +194,161 @@ pub fn is_strongly_connected(graph: &digraph::PyDiGraph) -> PyResult<bool> {
+    let mut node_map = vec![usize::MAX; g.node_count()];
+    for comp in components_usize.iter() {
+        let new_nix = condensed.add_node(Vec::new());
+        for nix in comp {
+            node_map[nix.index()] = new_nix.index();
+        }
+    }
+
+    // Consume nodes and edges of the old graph and insert them into the new one.
+    let (nodes, edges) = g.into_nodes_edges();
+    for (nix, node) in nodes.into_iter().enumerate() {
+        let idx = node_map.get(nix).copied().unwrap_or(usize::MAX);
+        if idx != usize::MAX {
+            condensed[NodeIndex::new(idx)].push(node.weight);
+        }
+    }
+    for edge in edges {
+        let source = node_map
+            .get(edge.source().index())
+            .copied()
+            .unwrap_or(usize::MAX);
+        let target = node_map
+            .get(edge.target().index())
+            .copied()
+            .unwrap_or(usize::MAX);
+        if source == usize::MAX || target == usize::MAX {
+            continue;
+        }
+        let source = NodeIndex::new(source);
+        let target = NodeIndex::new(target);
+        if make_acyclic && Ty::is_directed() {
+            if source != target {
+                condensed.update_edge(source, target, edge.weight);
+            }
+        } else {
+            condensed.add_edge(source, target, edge.weight);
+        }
+    }
+
+    let mapped = condensed.map(
+        |_, w| match w.clone().into_pyobject(py) {
+            Ok(bound) => bound.unbind(),
+            Err(_) => PyValueError::new_err("Node conversion failed")
+                .into_pyobject(py)
+                .unwrap()
+                .unbind()
+                .into(),
+        },
+        |_, w| match w.clone().into_pyobject(py) {
+            Ok(bound) => bound.unbind(),
+            Err(_) => PyValueError::new_err("Edge conversion failed")
+                .into_pyobject(py)
+                .unwrap()
+                .unbind()
+                .into(),
+        },
+    );
+    Ok((mapped, node_map))
+}
+
+#[pyfunction]
+#[pyo3(text_signature = "(graph, /, sccs=None)", signature=(graph, sccs=None))]
+pub fn condensation(
+    py: Python,
+    graph: PyObject,
+    sccs: Option<Vec<Vec<usize>>>,
+) -> PyResult<PyObject> {
+    if let Ok(digraph) = graph.extract::<digraph::PyDiGraph>(py) {
+        let g = digraph.graph.clone();
+        let (condensed, node_map) = condensation_inner(py, g.into(), true, sccs)?;
+        let mut result = digraph::PyDiGraph {
+            graph: condensed,
+            cycle_state: algo::DfsSpace::default(),
+            check_cycle: false,
+            node_removed: false,
+            multigraph: true,
+            attrs: PyDict::new(py).into(),
+        };
+        let node_map_py = node_map.into_pyobject(py)?;
+        let attrs = PyDict::new(py);
+        attrs.set_item("node_map", node_map_py)?;
+        result.attrs = attrs.into();
+        Ok(result.into_pyobject(py)?.into())
```

Comment:
If you split the Rust methods in `digraph_condensation` and `graph_condensation`, you'll be able to just return the `PyDiGraph` or `PyGraph` object without needing this conversion